circuits/aes-gcm/gctr.circom

```diff
@@ -0,0 +1,122 @@
+pragma circom 2.1.9;
+include "../aes-ctr/cipher.circom";
+include "utils.circom";
+include "../aes-ctr/ctr.circom";
+include "helper_functions.circom";
+// GCTR Process to be used in AES-GCM
+//
+//            ┌───────────┐           inc           ┌───────────┐
+//            │    ICB    │ ──────────────────────▶ │    CB2    │
+//            └───────────┘                         └───────────┘
+//                  │                                      │
+//                  ▼                                      ▼
+//            ┌──────────┐                           ┌──────────┐
+//            │  CIPH_K  │                           │  CIPH_K  │
+//            └──────────┘                           └──────────┘
+//                  │                                      │
+//                  ▼                                      ▼
+//             ┌──────────┐                           ┌──────────┐
+//             │    X1    │                           │     X2   │
+//             └──────────┘                           └──────────┘
+//                   │                                      │
+//   ┌───────┐       ▼                       ┌───────┐      ▼
+//   |  X_1  |───▶  XOR                      |  X_2  |───▶ XOR
+//   └───────┘       │                       └───────┘      │
+//                   ▼                                      ▼
+//               ┌──────┐                                ┌──────┐
+//               │  Y1  │                                │  Y2  │
+//               └──────┘                                └──────┘
+//                   │                                       │
+//                   ▼                                       ▼
+//
+// GCTR_K (ICB, X1 || X2 || ... || X_n*) = Y1 || Y2 || ... || Y_n*.
+
+
+template GCTR(INPUT_LEN, nk) {
+    signal input key[nk * 4];
+    signal input initialCounterBlock[128];
+    signal input plainText[INPUT_LEN];
+    signal output cipherText[INPUT_LEN];
+
+    // number of 128 bit blocks in the plaintext
+    var nBlocks = INPUT_LEN / 128;
+    // size of the last block
+    var lastBlockSize = INPUT_LEN % 128;
+    // total number of bits in the plaintext blocks
+    var bitblocks = 128 * nBlocks;
+
+    // last block of plaintext
+    signal tempLastBlock[lastBlockSize];
+    for (var i = 0; i < lastBlockSize -1; i++) {
+        tempLastBlock[i] <== plainText[nBlocks * 128 + i];
+    }
+
+    // generate plaintext blocks
+    // note to not use the last block of plaintext
+    // because it will be padded byt the toBlocks components
+    component plainTextBlocks = ToBlocks(INPUT_LEN);
```

Comment:
Instead of moving between blocks and stream in circuit, would it make more sense to accept input as sequence of blocks?